src/permission/fs_permission.cc

```diff
@@ -154,6 +155,20 @@ bool FSPermission::is_granted(PermissionScope perm,
   }
 }
 
+std::string FSPermission::RadixTree::NormalizePathIfCaseInsensitive(
+    const std::string& path) {
+  if (!case_sensitive_) {
+    std::string transformed_path = path;
+    std::transform(
+        transformed_path.begin(),
+        transformed_path.end(),
+        transformed_path.begin(),
+        [](unsigned char c) -> unsigned char { return std::tolower(c); });
```

Comment:
The current code copies the string and transforms the new copied string without vectorizing it, which results in unnecessary operations and performance degregations. There are couple of optimizations available in this piece of code, that we can leverage it in our benefit, since this function may be called a lot:

- Avoid using `std::tolower`:

`std::tolower` is really slow. We can use something like this to lowercase a string.

```
constexpr char to_lower(char x) noexcept { return (x | 0x20); }
```

- Avoid using `std::transform`

The following code is from `Ada` and provides auto-vectorizable lowercasing faster than `std::transform`

```c++
constexpr bool to_lowercase(char* input, size_t length) noexcept {
  auto broadcast = [](uint8_t v) -> uint64_t { return 0x101010101010101 * v; };
  uint64_t broadcast_80 = broadcast(0x80);
  uint64_t broadcast_Ap = broadcast(128 - 'A');
  uint64_t broadcast_Zp = broadcast(128 - 'Z');
  uint64_t non_ascii = 0;
  size_t i = 0;

  for (; i + 7 < length; i += 8) {
    uint64_t word{};
    memcpy(&word, input + i, sizeof(word));
    non_ascii |= (word & broadcast_80);
    word ^=
        (((word + broadcast_Ap) ^ (word + broadcast_Zp)) & broadcast_80) >> 2;
    memcpy(input + i, &word, sizeof(word));
  }
  if (i < length) {
    uint64_t word{};
    memcpy(&word, input + i, length - i);
    non_ascii |= (word & broadcast_80);
    word ^=
        (((word + broadcast_Ap) ^ (word + broadcast_Zp)) & broadcast_80) >> 2;
    memcpy(input + i, &word, length - i);
  }
  return non_ascii == 0;
}
```

Reply:
This code isn't being used in a hot path. Only in the `insert`. The `lookup` method uses a slightly different approach (without creating a new copy). So... about exactly how much improvement are we talking about? Because the above snippet sounds quite complex to maintain if it will bring a really small performance improvement.

Reply:
Note that, this function will be called only when `process.permission.deny` is called.

Reply:
Small footnote: `to_lowercase` is 5x faster - https://quick-bench.com/q/nAH_54VJi1--eP2ZM2S5sIXg54Q